src/transactions/InvokeHostFunctionOpFrame.cpp

```diff
@@ -250,23 +268,42 @@ InvokeHostFunctionOpFrame::doApply(Application& app, AbstractLedgerTxn& ltx)
     auto const& footprint = resources.footprint;
     ledgerEntryCxxBufs.reserve(footprint.readOnly.size() +
                                footprint.readWrite.size());
-    auto addReads = [&ledgerEntryCxxBufs, &ltx, &metrics](auto const& keys) {
+    auto addReads = [&ledgerEntryCxxBufs, &ltx, &metrics,
+                     &sorobanConfig](auto const& keys) -> bool {
         for (auto const& lk : keys)
         {
+            if (sorobanConfig.maxContractDataKeySizeBytes() < toVec(lk).size())
+            {
+                return true;
+            }
             // Load without record for readOnly to avoid writing them later
             auto ltxe = ltx.loadWithoutRecord(lk);
             size_t nByte{0};
             if (ltxe)
             {
-                auto buf = toCxxBuf(ltxe.current());
+                auto const& le = ltxe.current();
+                auto buf = toCxxBuf(le);
                 nByte = buf.data->size();
+                // These limits should have also been enforced during preflight
```

Comment:
I get the point about not requiring preflight. But here we can read a too large entry because the entry size limit have been reduced and someone is still trying to retrieve it, so it's not a logic error right?

Reply:
I'm actually not sure what should be the semantics here; downgrading the max entry size seems like a logic error to me (we might even implement the respective upgrade validation checks...). Probably it's ok to return the limit exceeded error here, just update a comment to not mention preflight and to mention that this should not normally happen unless some non-backwards-compatible upgrades happened.

Reply:
Updated the comments.